src/drivers/sdr/gnuradio.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2021  University of Alberta
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+/**
+ * @file gnuradio.c
+ * @author Josh Lazaruk
+ * @date 2022-05-12
+ */
+
+#include <assert.h>
+#include <csp/csp.h>
+#include <csp/drivers/usart.h>
+#include <csp/arch/csp_malloc.h>
+#include <csp/arch/csp_thread.h>
+
+#include <zmq.h>
+
+#ifdef CSP_POSIX
+#include <stdio.h>
+#define ex2_log printf
+#endif // CSP_POSIX
+
+#define SDR_GNURADIO
+#ifdef SDR_GNURADIO
+
+static void *context;
+static void *publisher;
+static void *subscriber;
+
+uint16_t crc16(char* pData, int length)
+{
+    uint8_t i;
+    uint16_t wCrc = 0xffff;
+    while (length--) {
+        wCrc ^= *(unsigned char *)pData++ << 8;
+        for (i=0; i < 8; i++)
+            wCrc = wCrc & 0x8000 ? (wCrc << 1) ^ 0x1021 : wCrc << 1;
+    }
+    return wCrc & 0xffff;
+}
+
+int csp_sdr_tx(const csp_route_t *ifroute, csp_packet_t *packet) {
+    csp_iface_t *iface = (csp_iface_t *)ifroute->iface;
+    csp_sdr_interface_data_t *ifdata = (csp_sdr_interface_data_t *)ifroute->iface->interface_data;
+    csp_sdr_conf_t *sdr_conf = (csp_sdr_conf_t *)iface->driver_data;
+
+    if (fec_csp_to_mpdu(ifdata->mac_data, packet, sdr_conf->mtu)) {
+        uint8_t *buf;
+        int delay_time = sdr_uhf_baud_rate_delay[ifdata->uhf_baudrate];
+        size_t mtu = (size_t)fec_get_next_mpdu(ifdata->mac_data, (void **)&buf);
+        while (mtu != 0) {
+            (ifdata->tx_func)(ifdata->fd, buf, mtu);
+            mtu = fec_get_next_mpdu(ifdata->mac_data, (void **)&buf);
+            csp_sleep_ms(delay_time);
+        }
+    }
+    return CSP_ERR_NONE;
+}
+
+CSP_DEFINE_TASK(csp_gnuradio_rx_task) {
+
+    usart_context_t * ctx = arg;
+
+    uint8_t rxframe[128] = {0};
```

Comment:
The 128 byte mtu is transferred to the code in this file through the sdr configuration. The configuration should be passed here so that the mtu can be taken from that configuration. Malloc or a VLA will need to be used (use malloc!!)